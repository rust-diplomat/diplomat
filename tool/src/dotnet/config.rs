use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct LibraryConfig {
    pub namespace: String,
    pub native_lib: String,
    #[serde(default)]
    pub usings: Vec<String>,
    #[serde(default)]
    pub exceptions: ExceptionsConfig,
    #[serde(default)]
    pub properties: PropertiesConfig,
}

impl LibraryConfig {
    pub fn default() -> LibraryConfig {
        LibraryConfig {
            namespace: "Interop".to_owned(),
            native_lib: "rust".to_owned(),
            usings: Vec::new(),
            exceptions: ExceptionsConfig::default(),
            properties: PropertiesConfig::default(),
        }
    }
}

#[derive(Debug, Deserialize, Default)]
pub struct PropertiesConfig {
    #[serde(default)]
    pub setters_prefix: Option<String>,
    #[serde(default)]
    pub getters_prefix: Option<String>,
}

#[derive(Debug, Deserialize, Default)]
pub struct ExceptionsConfig {
    #[serde(default)]
    pub trim_suffix: String,
    #[serde(default)]
    pub error_message_method: Option<String>,
}
