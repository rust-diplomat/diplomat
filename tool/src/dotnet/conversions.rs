use std::fmt::{self, Write as _};

use diplomat_core::{ast, Env};

use super::types::{gen_type_name_to_string, type_name_for_prim};
use crate::util::CodeWriter;

pub struct SliceParam {
    pub array_var_name: String,
    pub ptr_var_name: String,
    pub length_var_name: String,
    pub underlying_type: ast::PrimitiveType,
}

impl SliceParam {
    pub fn new(name: String, underlying_type: ast::PrimitiveType) -> Self {
        let ptr_var_name = format!("{}Ptr", name);
        let length_var_name = format!("{}Length", name);
        Self {
            array_var_name: name,
            ptr_var_name,
            length_var_name,
            underlying_type,
        }
    }

    pub fn open_fixed_block(&self, out: &mut CodeWriter) -> fmt::Result {
        writeln!(
            out,
            "fixed ({}* {} = {})",
            type_name_for_prim(&self.underlying_type),
            self.ptr_var_name,
            self.array_var_name,
        )?;
        writeln!(out, "{{")?;
        out.indent();
        Ok(())
    }

    pub fn close_fixed_block(self, out: &mut CodeWriter) -> fmt::Result {
        out.dedent();
        writeln!(out, "}}")
    }
}

/// Writes the raw → idiomatic conversion code
///
/// Primitive type: nothing to do
/// Struct/opaque custom type: wrap pointer into a managed class
/// Enum: cast from raw enum
pub fn to_idiomatic_object<W: fmt::Write>(
    env: &Env,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    input_var_name: &str,
    out: &mut W,
) -> fmt::Result {
    match typ {
        ast::TypeName::Primitive(_) => out.write_str(input_var_name),
        ast::TypeName::Box(boxed) => {
            to_idiomatic_object(env, boxed.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Reference(.., reference) => {
            to_idiomatic_object(env, reference.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Option(opt) => {
            to_idiomatic_object(env, opt.as_ref(), in_path, input_var_name, out)
        }
        _ => {
            let name = gen_type_name_to_string(typ, in_path, env)?;
            match typ {
                ast::TypeName::Named(path_type) => match path_type.resolve(in_path, env) {
                    ast::CustomType::Struct(_) | ast::CustomType::Opaque(_) => {
                        write!(out, "new {name}({input_var_name})")
                    }
                    ast::CustomType::Enum(_) => {
                        write!(out, "({name}){input_var_name}")
                    }
                },
                other => panic!("expected named type name, found `{}`", other),
            }
        }
    }
}

/// Writes the idiomatic → raw conversion code
///
/// Primitive type: nothing to do
/// Struct/opaque custom type: extract raw representation using `.AsFFI()`
/// Enum: cast from idiomatic enum
pub fn to_raw_object<W: fmt::Write>(
    env: &Env,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    input_var_name: &str,
    out: &mut W,
) -> fmt::Result {
    match typ {
        ast::TypeName::Primitive(_) => out.write_str(input_var_name),
        ast::TypeName::Box(boxed) => {
            to_raw_object(env, boxed.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Reference(.., reference) => {
            to_raw_object(env, reference.as_ref(), in_path, input_var_name, out)
        }
        ast::TypeName::Option(opt) => {
            to_raw_object(env, opt.as_ref(), in_path, input_var_name, out)
        }
        _ => {
            let name = gen_type_name_to_string(typ, in_path, env)?;
            match typ {
                ast::TypeName::Named(path_type) => match path_type.resolve(in_path, env) {
                    ast::CustomType::Struct(_) | ast::CustomType::Opaque(_) => {
                        write!(out, "{input_var_name}.AsFFI()")
                    }
                    ast::CustomType::Enum(_) => {
                        write!(out, "(Raw.{name}){input_var_name}")
                    }
                },
                other => panic!("expected named type name, found `{}`", other),
            }
        }
    }
}
