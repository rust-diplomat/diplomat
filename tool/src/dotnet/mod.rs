use std::collections::HashMap;
use std::fmt;
use std::fmt::Write;
use std::fs;
use std::path::PathBuf;

use colored::*;
use diplomat_core::Env;

use self::config::LibraryConfig;
use crate::util::CodeWriter;
use crate::util::SetOfAstTypes;

mod config;
mod conversions;
mod idiomatic;
mod raw;
mod types;
mod util;

static RUNTIME_CS: &str = include_str!("Runtime.cs");

const INDENTATION: &str = "    ";
const SCOPE_OPENING: &str = "{";
const SCOPE_CLOSING: &str = "}";

pub fn gen_bindings(
    env: &Env,
    library_config_path: &Option<PathBuf>,
    docs_url_gen: &diplomat_core::ast::DocsUrlGenerator,
    outs: &mut HashMap<String, String>,
) -> fmt::Result {
    let mut library_config = LibraryConfig::default();
    if let Some(path) = library_config_path {
        // Should be fine, we've already verified the path
        let contents = fs::read_to_string(path).unwrap();
        match toml::from_str(&contents) {
            Ok(config) => library_config = config,
            Err(err) => {
                eprintln!(
                    "{} Unable to parse library configuration file: {path:?}\n{err}",
                    "Error:".red().bold(),
                );
                std::process::exit(1);
            }
        }
    }

    let diplomat_runtime_out = outs
        .entry("DiplomatRuntime.cs".to_owned())
        .or_insert_with(String::new);
    write!(
        diplomat_runtime_out,
        "{}",
        RUNTIME_CS.replace("{{NAMESPACE}}", &library_config.namespace)
    )?;

    let all_types = crate::util::get_all_custom_types(env);
    let mut errors = SetOfAstTypes::default();

    {
        // Raw API generation pass

        let mut results = SetOfAstTypes::default();

        for (in_path, typ) in &all_types {
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            raw::gen_header(&library_config, &mut out)?;
            raw::gen(
                env,
                &library_config,
                &mut results,
                &mut errors,
                typ,
                in_path,
                docs_url_gen,
                &mut out,
            )?;
            outs.insert(format!("Raw{}.cs", typ.name()), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: Raw{}.cs", typ.name()));
        }

        for (in_path, typ) in results {
            let result_name = types::gen_type_name_to_string(typ, &in_path, env)?;
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            raw::gen_header(&library_config, &mut out)?;
            raw::gen_result(typ, &in_path, env, &mut out)?;
            outs.insert(format!("Raw{}.cs", result_name), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: Raw{}.cs", result_name));
        }
    }

    {
        // Idiomatic API generation pass

        for (in_path, typ) in &all_types {
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            idiomatic::gen_header(&library_config, &mut out)?;
            idiomatic::gen(typ, in_path, env, &library_config, docs_url_gen, &mut out)?;
            outs.insert(format!("{}.cs", typ.name()), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: {}.cs", typ.name()));
        }

        for (in_path, typ) in errors {
            let idiomatic::ExceptionCtx { name, .. } =
                idiomatic::error_type_to_exception_name(env, &library_config, typ, &in_path)?;
            let mut out_buf = String::new();
            let mut out = CodeWriter::new(&mut out_buf, INDENTATION, SCOPE_OPENING, SCOPE_CLOSING);
            idiomatic::gen_header(&library_config, &mut out)?;
            idiomatic::gen_exception(env, &library_config, typ, &in_path, &mut out)?;
            outs.insert(format!("{}.cs", name), out_buf)
                .and_then::<String, _>(|_| panic!("file created twice: {}.cs", name));
        }
    }

    Ok(())
}
