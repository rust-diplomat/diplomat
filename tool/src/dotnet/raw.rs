use std::fmt;
use std::fmt::Write as _;

use diplomat_core::ast;
use diplomat_core::Env;
use heck::ToLowerCamelCase;
use heck::ToUpperCamelCase;

use super::config::LibraryConfig;
use super::types::gen_type_name;
use super::util::{collect_errors, collect_results, gen_doc_block};
use crate::util::{CodeWriter, SetOfAstTypes};

pub fn gen_header(library_config: &LibraryConfig, out: &mut CodeWriter) -> fmt::Result {
    writeln!(out, "// <auto-generated/> by Diplomat")?;
    writeln!(out)?;

    writeln!(out, "#pragma warning disable 0105")?;
    writeln!(out, "using System;")?;
    writeln!(out, "using System.Runtime.InteropServices;")?;
    writeln!(out)?;
    for using in &library_config.usings {
        writeln!(out, "using {};", using)?;
    }
    writeln!(out, "using {}.Diplomat;", library_config.namespace)?;
    writeln!(out, "#pragma warning restore 0105")?;
    writeln!(out)?;

    writeln!(out, "namespace {}.Raw;", library_config.namespace)?;

    writeln!(out)?;
    writeln!(out, "#nullable enable")?;

    Ok(())
}

#[allow(clippy::too_many_arguments)]
pub fn gen<'ast>(
    env: &Env,
    library_config: &LibraryConfig,
    results: &mut SetOfAstTypes<&'ast ast::TypeName>,
    errors: &mut SetOfAstTypes<&'ast ast::TypeName>,
    typ: &'ast ast::CustomType,
    in_path: &'ast ast::Path,
    docs_url_gen: &ast::DocsUrlGenerator,
    out: &mut CodeWriter,
) -> fmt::Result {
    for method in typ.methods() {
        for param in &method.params {
            collect_results(&param.ty, in_path, env, results);
            collect_errors(&param.ty, in_path, env, errors);
        }

        if let Some(return_type) = method.return_type.as_ref() {
            collect_results(return_type, in_path, env, results);
            collect_errors(return_type, in_path, env, errors);
        }
    }

    writeln!(out)?;

    match typ {
        ast::CustomType::Struct(strct) => {
            for (_, typ, _) in &strct.fields {
                collect_results(typ, in_path, env, results);
                collect_errors(typ, in_path, env, errors);
            }

            gen_doc_block(out, &strct.docs.to_markdown(docs_url_gen))?;
            writeln!(out, "[StructLayout(LayoutKind.Sequential)]")?;
            writeln!(out, "public partial struct {}", typ.name())?;

            out.scope(|out| {
                writeln!(
                    out,
                    "private const string NativeLib = \"{}\";",
                    library_config.native_lib
                )?;

                for (name, typ, doc) in strct.fields.iter() {
                    gen_field(name, doc, typ, in_path, env, docs_url_gen, out)?;
                }

                for method in typ.methods() {
                    gen_method(typ, method, in_path, env, docs_url_gen, out)?;
                }

                Ok(())
            })
        }

        ast::CustomType::Opaque(opaque) => {
            gen_doc_block(out, &opaque.docs.to_markdown(docs_url_gen))?;
            writeln!(out, "[StructLayout(LayoutKind.Sequential)]")?;
            writeln!(out, "public partial struct {}", typ.name())?;

            out.scope(|out| {
                writeln!(
                    out,
                    "private const string NativeLib = \"{}\";",
                    library_config.native_lib
                )?;

                for method in typ.methods() {
                    gen_method(typ, method, in_path, env, docs_url_gen, out)?;
                }

                writeln!(out)?;
                writeln!(
                    out,
                    r#"[DllImport(NativeLib, CallingConvention = CallingConvention.Cdecl, EntryPoint = "{}_destroy", ExactSpelling = true)]"#,
                    typ.name()
                )?;
                writeln!(out, "public static unsafe extern void Destroy({}* self);", typ.name())
            })
        }

        ast::CustomType::Enum(enm) => {
            gen_doc_block(out, &enm.docs.to_markdown(docs_url_gen))?;
            writeln!(out, "public enum {}", enm.name)?;
            out.scope(|out| {
                for (name, discriminant, docs) in enm.variants.iter() {
                    gen_doc_block(out, &docs.to_markdown(docs_url_gen))?;
                    writeln!(out, "{} = {},", name, discriminant)?;
                }

                Ok(())
            })
        }
    }
}

fn gen_field(
    name: &ast::Ident,
    docs: &ast::Docs,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    docs_url_gen: &ast::DocsUrlGenerator,
    out: &mut CodeWriter,
) -> fmt::Result {
    let mut type_declaration = String::new();
    gen_type_name_decl_position(typ, in_path, env, &mut type_declaration)?;
    let is_unsafe = type_declaration.ends_with('*');

    writeln!(out)?;
    gen_doc_block(out, &docs.to_markdown(docs_url_gen))?;
    gen_annotations_for_field(typ, out)?;
    write!(out, "public ")?;
    if is_unsafe {
        write!(out, "unsafe ")?;
    }
    writeln!(out, "{type_declaration} {name};")
}

fn gen_method(
    typ: &ast::CustomType,
    method: &ast::Method,
    in_path: &ast::Path,
    env: &Env,
    docs_url_gen: &ast::DocsUrlGenerator,
    out: &mut CodeWriter,
) -> fmt::Result {
    writeln!(out)?;

    gen_doc_block(out, &method.docs.to_markdown(docs_url_gen))?;
    gen_annotations_for_method(method, out)?;
    write!(out, "public static unsafe extern ")?;
    gen_type_name_return_position(method.return_type.as_ref(), in_path, env, out)?;

    write!(
        out,
        " {}(",
        method
            .full_path_name
            .as_str()
            .replace(&format!("{}_", typ.name()), "")
            .to_upper_camel_case()
    )?;

    let mut first = true;

    if let Some(ref self_param) = method.self_param {
        gen_param("self", &self_param.to_typename(), false, in_path, env, out)?;
        first = false;
    }

    for param in method.params.iter() {
        if first {
            first = false;
        } else {
            write!(out, ", ")?;
        }

        let name = param.name.as_str().to_lower_camel_case();
        gen_param(&name, &param.ty, param.is_writeable(), in_path, env, out)?;
    }

    writeln!(out, ");")?;

    Ok(())
}

pub fn gen_result(
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    out: &mut CodeWriter,
) -> fmt::Result {
    let (ok, err) = if let ast::TypeName::Result(ok, err) = typ {
        (ok, err)
    } else {
        panic!("not a result: {:?}", typ);
    };

    writeln!(out)?;
    writeln!(out, "[StructLayout(LayoutKind.Sequential)]")?;
    write!(out, "public partial struct ")?;
    gen_type_name(typ, in_path, env, out)?;
    writeln!(out)?;

    out.scope(|out| {
        // Omit variants or even the entire union if parts are zero-sized.
        // This matches what rustc effectively does with zero-sized union variants
        if !ok.is_zst() || !err.is_zst() {
            writeln!(out, "[StructLayout(LayoutKind.Explicit)]")?;
            writeln!(out, "private unsafe struct InnerUnion")?;

            out.scope(|out| {
                if !ok.is_zst() {
                    writeln!(out, "[FieldOffset(0)]")?;
                    write!(out, "internal ")?;
                    gen_type_name_decl_position(ok, in_path, env, out)?;
                    writeln!(out, " ok;")?;
                }

                if !err.is_zst() {
                    writeln!(out, "[FieldOffset(0)]")?;
                    write!(out, "internal ")?;
                    gen_type_name_decl_position(err, in_path, env, out)?;
                    writeln!(out, " err;")?;
                }

                Ok(())
            })?;

            writeln!(out)?;
            writeln!(out, "private InnerUnion _inner;")?;
            writeln!(out)?;
        }

        writeln!(out, "[MarshalAs(UnmanagedType.U1)]")?;
        writeln!(out, "public bool isOk;")?;

        if !ok.is_zst() {
            writeln!(out)?;
            write!(out, "public unsafe ")?;
            gen_type_name_decl_position(ok, in_path, env, out)?;
            writeln!(out, " Ok")?;
            out.scope(|out| {
                writeln!(out, "get")?;
                out.scope(|out| writeln!(out, "return _inner.ok;"))
            })?;
        }

        if !err.is_zst() {
            writeln!(out)?;
            write!(out, "public unsafe ")?;
            gen_type_name_decl_position(err, in_path, env, out)?;
            writeln!(out, " Err")?;
            out.scope(|out| {
                writeln!(out, "get")?;
                out.scope(|out| writeln!(out, "return _inner.err;"))
            })?;
        }

        Ok(())
    })?;

    Ok(())
}

fn gen_annotations_for_method(method: &ast::Method, out: &mut dyn fmt::Write) -> fmt::Result {
    writeln!(
        out,
        r#"[DllImport(NativeLib, CallingConvention = CallingConvention.Cdecl, EntryPoint = "{}", ExactSpelling = true)]"#,
        method.full_path_name
    )?;
    match &method.return_type {
        Some(ast::TypeName::Primitive(ast::PrimitiveType::bool)) => {
            writeln!(out, "[return: MarshalAs(UnmanagedType.U1)]")
        }
        _ => Ok(()),
    }
}

fn gen_annotations_for_param(typ: &ast::TypeName, out: &mut dyn fmt::Write) -> fmt::Result {
    match typ {
        ast::TypeName::Primitive(ast::PrimitiveType::bool) => {
            write!(out, "[MarshalAs(UnmanagedType.U1)] ")
        }
        _ => Ok(()),
    }
}

fn gen_annotations_for_field(typ: &ast::TypeName, out: &mut dyn fmt::Write) -> fmt::Result {
    match typ {
        ast::TypeName::Primitive(ast::PrimitiveType::bool) => {
            writeln!(out, "[MarshalAs(UnmanagedType.U1)]")
        }
        _ => Ok(()),
    }
}

fn gen_type_name_decl_position(
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    out: &mut dyn fmt::Write,
) -> fmt::Result {
    match typ {
        ast::TypeName::Option(opt) => match opt.as_ref() {
            ast::TypeName::Box(ptr) | ast::TypeName::Reference(.., ptr) => {
                gen_type_name_decl_position(ptr.as_ref(), in_path, env, out)?;
                write!(out, "*")
            }
            _ => panic!("Options without a pointer type are not yet supported"),
        },
        ast::TypeName::Box(underlying) | ast::TypeName::Reference(.., underlying) => {
            gen_type_name_decl_position(underlying.as_ref(), in_path, env, out)?;
            write!(out, "*")
        }
        ast::TypeName::Unit => panic!("unexpected unit type in declaration position"),
        _ => gen_type_name(typ, in_path, env, out),
    }
}

fn gen_type_name_return_position<'ast>(
    typ: impl Into<Option<&'ast ast::TypeName>>,
    in_path: &ast::Path,
    env: &Env,
    out: &mut dyn fmt::Write,
) -> fmt::Result {
    match &typ.into() {
        None | Some(ast::TypeName::Unit) => write!(out, "void"),
        Some(other) => gen_type_name_decl_position(other, in_path, env, out),
    }
}

fn gen_param(
    name: &str,
    typ: &ast::TypeName,
    is_writeable: bool,
    in_path: &ast::Path,
    env: &Env,
    out: &mut dyn fmt::Write,
) -> fmt::Result {
    if is_writeable {
        write!(out, "DiplomatWriteable* {name}")
    } else {
        match typ {
            ast::TypeName::StrReference(..) => {
                write!(out, "byte* {name}, nuint {name}Sz")
            }
            ast::TypeName::PrimitiveSlice(.., prim) => {
                write!(
                    out,
                    "{}* {name}, nuint {name}Sz",
                    super::types::type_name_for_prim(prim),
                )
            }
            ast::TypeName::Option(opt) => gen_param(name, opt.as_ref(), false, in_path, env, out),
            _ => {
                gen_annotations_for_param(typ, out)?;
                gen_type_name_decl_position(typ, in_path, env, out)?;
                write!(out, " {name}")
            }
        }
    }
}
