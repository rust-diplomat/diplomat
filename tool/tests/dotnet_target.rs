use diplomat_tool::dotnet::*;

macro_rules! test_file {
    ($($file:tt)*) => {
        let parsed: syn::File = syn::parse_quote! { $($file)* };
        let custom_types = diplomat_core::ast::File::from(&parsed);
        let env = custom_types.all_types();

        let mut out_texts = std::collections::HashMap::new();

        gen_bindings(&env, None, &Default::default(), &mut out_texts).unwrap();
        out_texts.remove("DiplomatRuntime.cs");

        for out in out_texts.keys() {
            insta::with_settings!({ snapshot_suffix => out.clone() }, {
                insta::assert_snapshot!(out_texts.get(out).unwrap())
            });
        }
    }
}

macro_rules! test_file_using_library_config {
    ($($file:tt)*) => {
        let parsed: syn::File = syn::parse_quote! { $($file)* };
        let custom_types = diplomat_core::ast::File::from(&parsed);
        let env = custom_types.all_types();

        let mut out_texts = std::collections::HashMap::new();

        use std::path::PathBuf;
        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        path.push("tests/configs/dotnet_example.toml");
        gen_bindings(&env, Some(&path), &Default::default(), &mut out_texts).unwrap();
        out_texts.remove("DiplomatRuntime.cs");

        for out in out_texts.keys() {
            insta::with_settings!({ snapshot_suffix => out.clone() }, {
                insta::assert_snapshot!(out_texts.get(out).unwrap())
            });
        }
    }
}

#[test]
fn generation_using_default_config() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyLibError;

            impl MyLibError {
                pub fn to_display(out: &mut DiplomatWrite) {
                    unimplemented!()
                }
            }

            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn new_slice(v: &[f64]) -> Result<Box<MyStruct>, Box<MyLibError>> {
                    unimplemented!()
                }

                pub fn set_slice(&mut self, new_slice: &[f64]) -> Result<(), Box<MyLibError>> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn generation_using_library_config() {
    test_file_using_library_config! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyLibError;

            impl MyLibError {
                pub fn to_display(&self, out: &mut DiplomatWrite) {
                    unimplemented!()
                }
            }

            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn new_slice(v: &[f64]) -> Result<Box<MyStruct>, Box<MyLibError>> {
                    unimplemented!()
                }

                pub fn set_slice(&mut self, new_slice: &[f64]) -> Result<(), Box<MyLibError>> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn setters_getters_properties() {
    test_file_using_library_config! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct Foo;

            #[diplomat::opaque]
            struct Bar;

            impl Bar {
                pub fn get_foo(&self) -> Result<Box<Foo>, ()> {
                    unimplemented!()
                }

                pub fn set_foo(&mut self, foo: &Foo) -> Result<(), ()> {
                    unimplemented!()
                }

                pub fn get_name(&self, out: &mut DiplomatWrite) -> Result<(), ()> {
                    unimplemented!()
                }

                pub fn set_name(&mut self, new_name: &DiplomatStr) -> Result<(), ()> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn cross_module_struct_fields() {
    test_file! {
        #[diplomat::bridge]
        mod mod1 {
            use super::mod2::Bar;

            struct Foo {
                x: Bar,
            }
        }

        #[diplomat::bridge]
        mod mod2 {
            use super::mod1::Foo;

            struct Bar {
                y: Box<Foo>,
            }
        }
    }
}

#[test]
fn cross_module_struct_methods() {
    test_file! {
        #[diplomat::bridge]
        mod mod1 {
            use super::mod2::Bar;

            #[diplomat::opaque]
            struct Foo;

            impl Foo {
                pub fn to_bar(&self) -> Bar {
                    unimplemented!()
                }
            }
        }

        #[diplomat::bridge]
        mod mod2 {
            use super::mod1::Foo;

            struct Bar {
                y: Box<Foo>,
            }
        }
    }
}

#[test]
fn enum_documentation() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            /// Documentation for MyEnum.
            ///
            /// Some remarks about this struct.
            #[diplomat::rust_link(foo::MyEnum, Enum)]
            enum MyEnum {
                /// All about A.
                A,
                /// All about B.
                #[diplomat::rust_link(foo::MyEnum::B, EnumVariant)]
                B,
                /// All about C.
                ///
                /// Some remarks about this variant.
                ///
                /// Even more remarks.
                C
            }
        }
    }
}

#[test]
fn simple_non_opaque_struct() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            struct MyStruct {
                a: u8,
                b: u8,
            }

            impl MyStruct {
                pub fn new(a: u8, b: u8) -> MyStruct {
                    unimplemented!()
                }

                pub fn get_a(&self) -> u8 {
                    unimplemented!()
                }

                pub fn set_b(&mut self, b: u8) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn simple_opaque_struct() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn new(a: u8, b: u8) -> Box<MyStruct> {
                    unimplemented!()
                }

                pub fn get_a(&self) -> u8 {
                    unimplemented!()
                }

                pub fn set_b(&mut self, b: u8) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn method_taking_str() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn new_str(v: &DiplomatStr) -> Box<MyStruct> {
                    unimplemented!()
                }

                pub fn set_str(&mut self, new_str: &DiplomatStr) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn method_taking_slice() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn new_slice(v: &[f64]) -> Box<MyStruct> {
                    unimplemented!()
                }

                pub fn fill_slice(&self, v: &mut [f64]) {
                    unimplemented!()
                }

                pub fn set_slice(&mut self, new_slice: &[f64]) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn method_write_out() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct(UnknownType);

            impl MyStruct {
                pub fn write(&self, out: &mut DiplomatWrite) {
                    unimplemented!()
                }

                pub fn write_unit(&self, out: &mut DiplomatWrite) -> () {
                    unimplemented!()
                }

                pub fn write_result(&self, out: &mut DiplomatWrite) -> Result<(), u8> {
                    unimplemented!()
                }

                pub fn write_no_rearrange(&self, out: &mut DiplomatWrite) -> u8 {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn struct_documentation() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            /// Documentation for Foo.
            /// Second line.
            struct Foo {
                /// Documentation for x.
                x: u8,
            }

            impl Foo {
                /// Documentation for get_x.
                pub fn get_x(&self) -> u8 {
                    x
                }
            }
        }
    }
}

#[test]
fn enum_conversion() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            enum MyEnum {
                A, B, C
            }

            struct MyStruct {
                a: u8,
                b: MyEnum,
            }

            #[diplomat::opaque]
            struct Foo;

            impl Foo {
                pub fn get_struct(&self, a: u8, b: MyEnum) -> MyStruct {
                    MyStruct { a, b }
                }
            }
        }
    }
}

#[test]
fn option_conversion() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOtherStruct;

            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn create(my_other_struct: Option<&MyOtherStruct>) -> Option<Box<MyStruct>> {
                    unimplemented!();
                }
            }
        }
    }
}

#[test]
fn pointer_types() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOpaqueStruct(UnknownType);

            struct MyStruct<'a> {
                a: &'a MyOpaqueStruct,
            }

            impl<'a> MyStruct<'a> {
                pub fn new(foo: &'a MyOpaqueStruct, bar: &'a mut MyOpaqueStruct) -> MyStruct<'a> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn boolean_type() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            struct MyStruct {
                a: bool,
            }

            impl MyStruct {
                pub fn func(a: bool) -> bool {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn option_types() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOpaqueStruct(UnknownType);

            struct MyStruct {
                a: Option<Box<MyOpaqueStruct>>,
            }
        }
    }
}

#[test]
fn result_types() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyOpaqueStruct(UnknownType);

            struct MyStruct {
                a: DiplomatResult<Box<MyOpaqueStruct>, u8>,
            }

            impl MyStruct {
                pub fn new() -> Result<MyStruct, u8> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn string_reference() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn new(v: &DiplomatStr) -> MyStruct {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn write_out() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn write(&self, to: &mut DiplomatWrite) {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn unit_type() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                pub fn something(&self) -> () {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn error_handling() {
    test_file! {
        #[diplomat::bridge]
        mod ffi {
            #[diplomat::opaque]
            struct MyStruct;

            #[diplomat::opaque]
            struct MyModuleOpaqueError;

            enum MyModuleError {
                BadInput,
                BadLuck,
            }

            impl MyStruct {
                pub fn foo(&self) -> Result<u32, MyModuleError> {
                    unimplemented!()
                }

                pub fn bar(&self) -> Result<DiplomatChar, MyModuleOpaqueError> {
                    unimplemented!()
                }

                pub fn baz(&self) -> Result<(), ()> {
                    unimplemented!()
                }
            }
        }
    }
}

#[test]
fn almost_properties() {
    test_file_using_library_config! {
        #[diplomat::bridge]
        mod ffi {
            /// This should not contain any property
            #[diplomat::opaque]
            struct MyStruct;

            impl MyStruct {
                /// This should not generate a property
                pub fn get_foo_by_key(&self, key: &DiplomatStr) -> Result<u64, ()> {
                    unimplemented!()
                }

                /// This should not generate a property
                pub fn set_foo_by_key(&self, key: &DiplomatStr, foo: i64) -> Result<(), ()> {
                    unimplemented!()
                }

                /// This should not generate a property
                pub fn get_str_by_key(&self, key: &DiplomatStr, writer: &mut DiplomatWrite) {
                    unimplemented!()
                }

                /// This should not generate a property
                pub fn set_str_by_key(&self, key: &DiplomatStr, s: &DiplomatStr) {
                    unimplemented!()
                }
            }
        }
    }
}
