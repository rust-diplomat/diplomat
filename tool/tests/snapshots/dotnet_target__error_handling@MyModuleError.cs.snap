---
source: tool/tests/dotnet_target.rs
expression: out_texts.get(out).unwrap()

---
// Automatically generated by Diplomat

#pragma warning disable 0105
using System;
using System.Runtime.InteropServices;

using Interop.Diplomat;
#pragma warning restore 0105

namespace Interop;

#nullable enable

public enum MyModuleError
{
    BadInput = 0,
    BadLuck = 1,
}

